tmt/steps/prepare/feature.py

```diff
@@ -0,0 +1,356 @@
+import dataclasses
+import enum
+from typing import Optional
+
+import tmt
+import tmt.base
+import tmt.log
+import tmt.options
+import tmt.steps
+import tmt.steps.prepare
+import tmt.utils
+from tmt.steps.provision import Guest
+from tmt.utils import ShellScript, field
+
+
+class Distro(enum.Enum):
+    FEDORA = enum.auto()
+    RHEL_7 = enum.auto()
+    RHEL_8 = enum.auto()
+    RHEL_9 = enum.auto()
+    CENTOS_7 = enum.auto()
+    CENTOS_STREAM_8 = enum.auto()
+    CENTOS_STREAM_9 = enum.auto()
+
+
+class Feature(tmt.utils.Common):
+    """ Base class for feature implementations """
+
+    KEY: str
+
+    def __init__(
+            self,
+            *,
+            parent: 'PrepareFeature',
+            guest: Guest,
+            logger: tmt.log.Logger) -> None:
+        """ Initialize feature data """
+        super().__init__(logger=logger, parent=parent, relative_indent=0)
+        self.guest = guest
+        self.logger = logger
+        self.guest_sudo = self.get_guest_sudo()
+        self.guest_distro = self.get_guest_distro()
+
+    def get_guest_sudo(self) -> str:
+        """ Return 'sudo' if guest is not superuser """
+        if not self.guest.facts.is_superuser:
+            return 'sudo'
+        return ''
+
+    def get_guest_distro(self) -> Optional[Distro]:
+        """ Get guest distro by parsing the guest facts """
+        os_release = self.guest.facts.os_release_content
+        if os_release is None:
+            return None
+
+        if os_release.get('NAME') == 'Fedora Linux':
+            return Distro.FEDORA
+
+        if os_release.get('NAME') == 'Red Hat Enterprise Linux':
+            if os_release.get('VERSION_ID', '').startswith('9'):
+                return Distro.RHEL_9
+            if os_release.get('VERSION_ID', '').startswith('8'):
+                return Distro.RHEL_8
+            if os_release.get('VERSION_ID', '').startswith('7'):
+                return Distro.RHEL_7
+
+        if os_release.get('NAME') == 'CentOS Linux' and os_release.get(
+                'VERSION_ID', '').startswith('7'):
+            return Distro.CENTOS_7
+
+        if os_release.get('NAME') == 'CentOS Stream':
+            if os_release.get('VERSION_ID', '').startswith('8'):
+                return Distro.CENTOS_STREAM_8
+            if os_release.get('VERSION_ID', '').startswith('9'):
+                return Distro.CENTOS_STREAM_9
+
+        return None
+
+
+class ToggleableFeature(Feature):
+    def enable(self) -> None:
+        raise NotImplementedError
+
+    def disable(self) -> None:
+        raise NotImplementedError
+
+
+FEDORA_REPO = 'powertools'
+FEDORA_PACKAGES = ['epel-release', 'epel-next-release']
+
+CENTOS_7_PACKAGES = ['epel-release']
+
+CENTOS_STREAM_9_REPO = 'crb'
+
+RHEL_7_REPO = 'rhel-*-optional-rpms rhel-*-extras-rpms rhel-ha-for-rhel-*-server-rpms'
```

Comment:
Would it make sense now to start moving this out to some configuration file rather?
Something in `/etc/tmt/prepare/feature.conf` or something?

Seems to me most of the info hardcoded data should land in config, make the code generic, so we can easily then add RHEL-10.0 withouth any code changes really. If we omit this here, it will be anyway needed to be refactored later ...

Reply:
Hi @thrix, it looks nice to use configuration file. Currently I don't know how to load `Distro` from 

```yaml
distros:
    - FEDORA
    - RHEL_7
    - RHEL_8
    - RHEL_9
    - CENTOS_7
    - CENTOS_STREAM_8
    - CENTOS_STREAM_9
```

to

```python
class Distro(enum.Enum):
    FEDORA = enum.auto()
    RHEL_7 = enum.auto()
    RHEL_8 = enum.auto()
    RHEL_9 = enum.auto()
    CENTOS_7 = enum.auto()
    CENTOS_STREAM_8 = enum.auto()
    CENTOS_STREAM_9 = enum.auto()
```

@happz, any suggestion?

Reply:
It'd be more along the lines of "distro => enable/disable commands", but this could be quite complicated, basically your `get_guest_distro()` would probably land in the config file. Something like this:

```yaml
- when: os-release.NAME == "Fedora Linux"
  # no enable or disable, no-op

- when: os-release.NAME == 'Red Hat Enterprise Linux' and os_release.VERSION_ID =~ "^9"
  enable:
    - '{sudo} dnf -y install https://dl.fedoraproject.org/pub/epel/epel-release-latest-9.noarch.rpm'
  disable:
    - '{sudo} dnf -y remove epel-release'
```

You get the idea, right? We'd move the map of supported distros & commands to enable/disable EPEL (or CRB, in its own section, and Buildroot, in its own section, and so on). Then we wouldn't need an enum, our plugin would simply traverse the config file, find the (first) matching entry, and run the `enable` or `disable` scripts.

I agree, this would be awesome, but, we can do this later - once the user-facing API is established, we can change the source of those scripts, and implement a clever way of finding them in the config file, a config file we don't have yet :) I for one would not block on this, but I'd definitely file an issue for this as it's definitely a worthy goal - adding support for the Ubuntu Testing repo would be just a couple of config lines :)

Reply:
Hi @happz, thank you very much for your explanation and the sample code. I'm with you to do it later, thanks again! (CC @thrix)

Reply:
Covered by moving to using `ansible` for the task.